brainiak/isfc.py

```diff
@@ -1,226 +1,1104 @@
+            prng = random_state
+        else:
+            prng = np.random.RandomState(random_state)
+
+        # Randomly sample subject IDs with replacement
+        subject_sample = sorted(prng.choice(np.arange(n_subjects),
+                                                 size=n_subjects))
+        
+        # Squareform and shuffle rows/columns of pairwise ISC matrix to
+        # to retain correlation structure among ISCs, then get triangle
+        if pairwise:
+            
+            # Loop through voxels
+            isc_sample = []
+            for voxel_iscs in iscs.T:
+
+                # Square the triangle and fill diagonal
+                voxel_iscs = squareform(voxel_iscs)
+                np.fill_diagonal(voxel_iscs, 1)
+
+                # Check that pairwise ISC matrix is square and symmetric
+                assert voxel_iscs.shape[0] == voxel_iscs.shape[1]
+                assert np.allclose(voxel_iscs, voxel_iscs.T)
+
+                # Shuffle square correlation matrix and get triangle
+                voxel_sample = voxel_iscs[subject_sample, :][:, subject_sample]
+                voxel_sample = squareform(voxel_sample, checks=False)
+
+                # Censor off-diagonal 1s for same-subject pairs
+                voxel_sample[voxel_sample == 1.] = np.NaN
+
+                isc_sample.append(voxel_sample)
+                
+            isc_sample = np.column_stack(isc_sample)
+
+        # Get simple bootstrap sample if not pairwise
+        elif not pairwise:
+            isc_sample = iscs[subject_sample, :]
+            
+        # Compute summary statistic for bootstrap ISCs per voxel
+        # (alternatively could construct distrubtion for all voxels
+        # then compute statistics, but larger memory footprint)
+        if summary_statistic == np.mean:
+            distribution.append(np.tanh(np.nanmean(np.arctanh(isc_sample), axis=0)))
+        elif summary_statistic == np.median:
+            distribution.append(np.nanmedian(isc_sample, axis=0))
+                    
+        # Update random state for next iteration
+        random_state = np.random.RandomState(prng.randint(0, 2**32 - 1))
+        
+    # Convert distribution to numpy array
+    distribution = np.array(distribution)
+    assert distribution.shape == (n_bootstraps, n_voxels)
+
+    # Compute CIs of median from bootstrap distribution (default: 95%)
+    ci = (np.percentile(distribution, (100 - ci_percentile)/2, axis=0),
+          np.percentile(distribution, ci_percentile + (100 - ci_percentile)/2, axis=0))
+    
+    # Shift bootstrap distribution to 0 for hypothesis test
+    shifted = distribution - observed
+    
+    # Get p-value for actual median from shifted distribution
+    p = ((np.sum(np.abs(shifted) >= np.abs(observed), axis=0) + 1) /
+          float((len(shifted) + 1)))[np.newaxis, :]
+    
+    if return_distribution:
+        return observed, ci, p, distribution
+    elif not return_distribution:
```

Comment:
I kind of like this ... 
```
if P:
    do A
elif not P:  
    do B
```
... structure, compared to ... 
```
if P:
    do A
do B
``` 
since the 1st structure might be more readable for people without programming background. Though I was a little confused initially. I wonder how other people think about this? @mihaic @manojneuro @cbaldassano

Reply:
@qihongl, the two examples you give are not equivalent as written, when `do A` does not necessarily interrupt the control flow (e.g., it is not `return A`).

Also, if the condition tested is the same, we should not repeat it in the `else` clause:
```python
if P:
    do A
else:
    do B
```

I'm not sure I'm completely on topic, but I would say we should have as few returns as possible, generally.

Reply:
@mihaic Right, I agree. 

The main point I want to make here is that, there are about 20 instances of `if P ... elif not P`. I wonder if we should switch all of them to something more standard.

Reply:
Indeed. We should not repeat the condition if not necessary. I see no benefit to repeating the condition.